packages/utils/src/store/ControllableStore.ts

```diff
@@ -0,0 +1,82 @@
+import { useIsoLayoutEffect } from '../useIsoLayoutEffect';
+import { Store } from './Store';
+
+export class ControllableStore<State> extends Store<State> {
+  /**
+   * Keeps track of which properties are controlled.
+   */
+  #controlledValues: Map<keyof State, boolean> = new Map();
+
+  #changeCallbacks: Map<keyof State, (value: any, ...rest: any) => void> = new Map();
+
+  useProp<Key extends keyof State, Value extends State[Key]>(key: keyof State, value: Value) {
+    // False positive - ESLint thinks we're calling a hook from a class component.
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    useIsoLayoutEffect(() => {
+      if (this.state[key] !== value) {
+        this.set(key, value);
+      }
+    }, [key, value]);
+  }
+
+  useProps(props: Partial<State>) {
+    // False positive - ESLint thinks we're calling a hook from a class component.
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    useIsoLayoutEffect(() => {
+      this.apply(props);
+    }, [props]);
```

Comment:
So to avoid calling `.apply()` too frequently within the root component, `props` must be a memo'd object

Reply:
Not really. `apply` itself does change detection so if no properties change, the store won't be updated. Calling `useMemo` would just move this change detection elsewhere (and introduce an unnecessary allocation).